pageserver/src/walingest.rs

```diff
@@ -470,14 +470,14 @@ impl<'a> WalIngest<'a> {
                         // we can't validate the remaining number of bytes without parsing
                         // the tuple data.
                         if (xlrec.flags & pg_constants::XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED) != 0 {
-                            old_heap_blkno = Some(decoded.blocks[0].blkno);
+                            old_heap_blkno = Some(decoded.blocks[1].blkno);
                         }
                         if (xlrec.flags & pg_constants::XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED) != 0 {
                             // PostgreSQL only uses XLH_UPDATE_NEW_ALL_VISIBLE_CLEARED on a
                             // non-HOT update where the new tuple goes to different page than
                             // the old one. Otherwise, only XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED is
                             // set.
-                            new_heap_blkno = Some(decoded.blocks[1].blkno);
+                            new_heap_blkno = Some(decoded.blocks[0].blkno);
```

Comment:
Hmm, what happens on a same-page update now? Reading the code in heap_update(), I think if the old and new tuple are on same page:
- only block ref 0 is used
- XLH_UPDATE_OLD_ALL_VISIBLE_CLEARED is set.

Isn't the `old_heap_blkno = Some(decoded.blocks[1].blkno);` assignment wrong in that case?

I'm surprised none of the existing tests catch that.

Reply:
Expanded the test case to cover that case. Indeed, you now get an error:
```
2023-09-14T21:57:56.770817Z ERROR wal_connection_manager{tenant_id=34c2e198bfd42a89284d97c4aa26aeba timeline_id=613e7b5e2bab2a021d30addf2fea084f}:connection{node_id=1}:panic{thread=walreceiver worker location=pageserver/src/walingest.rs:529:51}: index out of bounds: the len is 1 but the index is 1
```